aptos-labs/aptos-core
# [forge] resurrect set image tag

### Description

Give Forge k8s the ability to upgrade validator StatefulSet image tags, which lets us write tests that roll out a new image tag upgrade across the cluster, while maintaining backwards compatibility.

#### Set image tags
Give Forge k8s the ability to upgrade validator StatefulSet image tags. This can be invoked via a a test or via the command line. For example, you can now do:

```
$ cargo run -p forge-cli -- operator set-node-image-tag \
  --image-tag 5da07eed1df564a0b28b1183a6ea32515b9af16e \
  --namespace forge-rustielin \
  --stateful-set-name aptos-node-28-validator \
  --container-name validator
```

Under the hood, we're just invoking `kubectl set image`, and then running a healthcheck against the node.

#### Compat test 

Also bring back the `compat` test, which runs with a lower number of validators (5), since this test is only for validating correctness. The test performs validator upgrades in this order, generating txn traffic between each step:
* spin up the network as version A
* upgrade 1 validator to version B
* upgrade half to version B
* upgrade rest to version B

Note that because network configuration is done outside of tests themselves, the `forge-cli` now has `--upgrade-image-tag` in addition to `--image-tag`

#### StatefulSet Cleanup

Also move a bunch of kube StatefulSet helper functions to a separate file `stateful_set.rs`, as `cluster_helper.rs` was getting bloated and hard to read. Improves the logs and errors in all the logic that waits on StatefulSets too. For example, we will now fail-fast if we see `ImagePullBackoff`, `CrashLoopBackOff`, etc.

### Test Plan

Run Forge locally:

```
$ IMAGE_TAG=daab3b690cd4a4f49bf73a13e85659d695c90f51 BASE_IMAGE_TAG=a8910fff6a9bd485fdeff12cfa6fd694e6ab61ba \
  FORGE_NAMESPACE=forge-rustielin FORGE_TEST_SUITE=compat \
  FORGE_NAMESPACE_KEEP=true FORGE_RUNNER_MODE=local ./testsuite/run_forge.sh
...
...
Compatibility test results for daab3b690cd4a4f49bf73a13e85659d695c90f51 ==> daab3b690cd4a4f49bf73a13e85659d695c90f51 (PR)
1. Check liveness of validators at old version: daab3b690cd4a4f49bf73a13e85659d695c90f51
compatibility::simple-validator-upgrade::liveness-check : 2577 TPS, 1087 ms latency, 1400 ms p99 latency,no expired txns
2. Upgrading first Validator to new version: daab3b690cd4a4f49bf73a13e85659d695c90f51
compatibility::simple-validator-upgrade::single-validator-upgrade : 2788 TPS, 1043 ms latency, 1050 ms p99 latency,no expired txns
3. Upgrading rest of first batch to new version: daab3b690cd4a4f49bf73a13e85659d695c90f51
compatibility::simple-validator-upgrade::half-validator-upgrade : 2785 TPS, 1218 ms latency, 3100 ms p99 latency,no expired txns
4. upgrading second batch to new version: daab3b690cd4a4f49bf73a13e85659d695c90f51
compatibility::simple-validator-upgrade::rest-validator-upgrade : 2836 TPS, 1037 ms latency, 1050 ms p99 latency,no expired txns
5. check swarm health
Compatibility test for daab3b690cd4a4f49bf73a13e85659d695c90f51 ==> daab3b690cd4a4f49bf73a13e85659d695c90f51 passed
```




- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/aptos-labs/aptos-core/2772)
